File: ja/codes/ruby/chapter_hashing/hash_map.rb
=begin
File: hash_map.rb
Created Time: 2024-04-14
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

require_relative '../utils/print_util'

### Driver Code ###
if __FILE__ == $0
  # ハッシュテーブルを初期化
  hmap = {}

  # 追加操作
  # ハッシュテーブルにキーと値の組 (key, value) を追加する
  hmap[12836] = "シャオハー"
  hmap[15937] = "シャオルオ"
  hmap[16750] = "シャオスワン"
  hmap[13276] = "シャオファー"
  hmap[10583] = "シャオヤー"
  puts "\n追加完了後、ハッシュテーブルは\nKey -> Value"
  print_hash_map(hmap)

  # 検索操作
  # ハッシュテーブルにキー key を入力し、値 value を取得する
  name = hmap[15937]
  puts "\n学籍番号 15937 を入力すると、名前 #{name} が見つかりました"

  # 削除操作
  # ハッシュテーブルからキーと値の組 (key, value) を削除する
  hmap.delete(10583)
  puts "\n10583 を削除した後、ハッシュテーブルは\nKey -> Value"
  print_hash_map(hmap)

  # ハッシュテーブルを走査
  puts "\nキーと値のペア Key->Value を走査"
  hmap.entries.each { |key, value| puts "#{key} -> #{value}" }

  puts "\nキー Key のみを個別に走査"
  hmap.keys.each { |key| puts key }

  puts "\n値 Value のみを個別に走査"
  hmap.values.each { |val| puts val }
end

File: ja/codes/typescript/chapter_hashing/hash_map.ts
/**
 * File: hash_map.ts
 * Created Time: 2022-12-20
 * Author: Daniel (better.sunjian@gmail.com)
 */

/* Driver Code */
/* ハッシュテーブルを初期化 */
const map = new Map<number, string>();

/* 追加操作 */
// ハッシュテーブルにキーと値のペア (key, value) を追加
map.set(12836, 'シャオハー');
map.set(15937, 'シャオルオ');
map.set(16750, 'シャオスワン');
map.set(13276, 'シャオファー');
map.set(10583, 'シャオヤー');
console.info('\n追加完了後，ハッシュテーブルは\nKey -> Value');
console.info(map);

/* 検索操作 */
// キー key をハッシュテーブルに渡し、値 value を取得
let name = map.get(15937);
console.info('\n学籍番号 15937 を入力し，見つかった氏名 ' + name);

/* 削除操作 */
// ハッシュテーブルからキーと値のペア (key, value) を削除
map.delete(10583);
console.info('\n10583 を削除した後，ハッシュテーブルは\nKey -> Value');
console.info(map);

/* ハッシュテーブルを走査 */
console.info('\nキーと値の組 Key->Value を走査');
for (const [k, v] of map.entries()) {
    console.info(k + ' -> ' + v);
}
console.info('\nキー Key を個別に走査');
for (const k of map.keys()) {
    console.info(k);
}
console.info('\n値 Value を個別に走査');
for (const v of map.values()) {
    console.info(v);
}

export {};
